src/storage/txn/process.rs

```diff
@@ -576,41 +586,73 @@ fn process_write_impl<S: Snapshot, L: LockManager>(
                 }
             }
             let mut txn = if scan_mode.is_some() {
-                MvccTxn::for_scan(snapshot, scan_mode, start_ts, !ctx.get_not_fill_cache())
+                MvccTxn::for_scan(
+                    snapshot,
+                    scan_mode,
+                    start_ts,
+                    !ctx.get_not_fill_cache(),
+                    pd_client,
+                )
             } else {
-                MvccTxn::new(snapshot, start_ts, !ctx.get_not_fill_cache())
+                MvccTxn::new(snapshot, start_ts, !ctx.get_not_fill_cache(), pd_client)
             };
 
             // Set extra op here for getting the write record when check write conflict in prewrite.
             txn.extra_op = extra_op;
 
+            let primary_key = Key::from_raw(&primary);
             let mut locks = vec![];
+            let mut async_commit_ts = TimeStamp::zero();
             for m in mutations {
+                let secondaries = if m.key() == &primary_key {
+                    &secondary_keys
+                } else {
+                    &None
+                };
                 match txn.prewrite(
                     m,
                     &primary,
+                    secondaries,
                     skip_constraint_check,
                     lock_ttl,
                     txn_size,
                     min_commit_ts,
                 ) {
-                    Ok(_) => {}
+                    Ok(ts) => {
+                        if secondaries.is_some() {
```

Comment:
should check `is_async_commit` instead of checking `secondaries`

Reply:
> When we are prewriting secondary keys

In this case secondary keys will be Some([]), not None.